beard/clustering/preclustering.py

```diff
@@ -0,0 +1,392 @@
+                first_names_left = len(first_names)
+                for reversed_index, name in \
+                        enumerate(reversed(tokenized_prefix)):
+                    if first_names_left == 0:
+                        return True
+                    elif first_names[-(reversed_index + 1)] != name:
+                        break
+                    first_names_left -= 1
+                    if reversed_index == len(tokenized_prefix) - 1:
+                        return True
+            return False
+        self._raise_keyerror(last_name)
+
+    def contains(self, tokens):
+        """Check if there is at least one signature with given surnames.
+
+        Parameters
+        ----------
+        :param tokens: tuple
+            Tokens which represent all surnames. Tuple of strings
+
+        Returns
+        -------
+        :returns: boolean
+            True if there is at least one sinature with given surnames.
+        """
+        return tokens in self._content
+
+    def initials_score(self, new_first_names, last_name):
+        """Count matches among the initials.
+
+        All first names from the new signature should be matched. A match
+        is full string match or a match between initials if at least one of the
+        compared string is only an initial. An initial is the first character
+        from the result of the double metaphone algorithm.
+
+        Two characters in initials - ``A`` and ``H`` are handled in special
+        way, as the double metaphone can output an empty string for them.
+
+        Parameters
+        ----------
+        :param new_first_names: tuple
+            Tuple of strings representing given names of the new author
+        :param last_name: tuple
+            Tokens, usually one, representing last surname(s) of the author.
+
+        Raises
+        ------
+        :raises: KeyError
+            When the last name is not included in the cluster
+
+        Returns
+        -------
+        :returns: integer
+            Score for matching initials in the cluster.
+        """
+        result = 0
+        if last_name in self._content:
+            for first_names, occurences in \
+                    six.iteritems(self._content[last_name]):
+                first_names_length = len(first_names)
+                old_names_index = 0
+                names_match = self.FULL_NAME_MATCH_BONUS
+                for initial in new_first_names:
+                    while old_names_index < first_names_length:
+                        full_names_match = \
+                            self._same_initials(first_names[old_names_index],
+                                                initial)
+                        if not full_names_match:
+                            old_names_index += 1
+                        else:
+                            if names_match == self.FULL_NAME_MATCH_BONUS:
+                                names_match = full_names_match
```

Comment:
This algorithm gives equal scores matching for examples 'Mark' to 'Mark' and to ' Mark Bell'. I propose dividing by the number of tokens to have better scoring